test_plans/execution_model.asciidoc

```diff
@@ -0,0 +1,112 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for execution model
+
+This is a test plan for SYCL command groups and execution order and SYCL application memory model as described in Section 3.7.1 and Section 3.8.1 of the SYCL 2020 specification.
+
+== Testing scope
+
+=== Backend coverage
+
+All the tests described below are performed for any SYCL implementation.
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+== Tests
+
+=== Execution order of three command groups submitted to the same queue
+
+Create 2 buffers for large arrays. Submit 2 commands that have some long loops and then fill these buffers with expected values via accessors in loop.
+Submit command that copy content of these buffers to result buffer.
+
+[source,c++]
+----
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc1(buffer1, cgh);
+  cgh.single_task([=] {
+    // long loop
+    for (int i = SIZE; i >= 0; i--) acc1[i] = i;
+  });
+});
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc2(buffer2, cgh);
+  cgh.single_task([=] {
+    // long loop
+    for (int i = SIZE; i >= 0; i--) acc2[i] = i;
+  });
+});
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc1(buffer1, cgh);
+  sycl::accessor<int, 1> acc2(buffer2, cgh);
+  sycl::accessor<int, 1> res_acc1(res_buffer1, cgh);
+  sycl::accessor<int, 1> res_acc2(res_buffer2, cgh);
+  cgh.parallel_for(sycl::range<1>(SIZE), [=](id<1> index) {
+    res_acc1[index] = acc1[index];
+    res_acc2[index] = acc2[index];
+  });
+});
+----
+
+Check that all elements in result buffer are equal to expected values which means that requisites for third command were satisfied before it started to execute.
+
+=== Execution order of three command groups submitted to the different queues
+
+Same as previous tests but each command is run on different queue with diffrent context.
+
+=== Requirements on overlapping sub-buffers
+
+This check is skipped if device doesn't support `aspect::usm_atomic_device_allocations`.
+
+Create buffer on array, and two overlapping sub-buffers
+
+[source,c++]
+----
+buffer<int, 1> sub_buf1(buffer, 0, 5);
+buffer<int, 1> sub_buf2(buffer, 3, 10);
+int* pflag = sycl::malloc_device<int>(1, queue);
+// assign pflag value to 0
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc1(sub_buf1, cgh);
+  cgh.single_task([=] {
+    sycl::atomic_ref<int> rflag(*pflag);
+    // long loop
+    for (int i = acc1.size(); i >= 0; i--) acc1[i] = i;
+    rflag = 1;
+  });
+});
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc2(sub_buf2, cgh);
+  sycl::accessor<bool, 1> res_acc(res_buffer, cgh);
+  cgh.single_task([=] {
+    res_acc[0] = *pflag == 1;
```

Comment:
I think the point of this test is to validate that the runtime waits for the first kernel to complete before executing the second kernel.  The logic for checking that seems broken.  It seems like the second kernel should use `atomic_ref` to set `*pflag` to `2`.  Then, the validation logic should verify that `*pflag == 2` after both kernels complete.

The current validation logic checks if `*pflag == 1`, but that will of course be true because it is only the first kernel that writes to `*pflag` and it sets the value to `1`.

Reply:
I see.  I think this description needs to be updated then:

> Check that *pflag was equal to 1, that means that requisites for second command with overlapping sub-buffer were satisfied before it started to execute.

To say something like:

> Check that `res_buffer` is equal to `true`, which means that requisites for second command with overlapping sub-buffer were satisfied before it started to execute.

Reply:
You should also use `atomic_ref` consistently.  If you want to use `atomic_ref` to set the value of `*pflag` in the first kernel, you should also use `atomic_ref` to read the value of `*pflag` in the second kernel.

Reply:
Thank you! Fixed.